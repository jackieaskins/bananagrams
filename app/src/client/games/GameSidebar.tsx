import {
  CloseButton,
  Flex,
  Heading,
  useColorModeValue,
} from "@chakra-ui/react";
import PlayerAccordion from "@/client/players/PlayerAccordion";

export type GameSidebarProps = {
  isOpen: boolean;
  onClose: () => void;
};

export default function GameSidebar({
  isOpen,
  onClose,
}: GameSidebarProps): JSX.Element {
  const borderColor = useColorModeValue("gray.400", "gray.700");

  return (
    <Flex
      height="100vh"
      width={isOpen ? "350px" : "0px"}
      transition="ease-in-out 0.2s"
      borderLeftWidth={1}
      borderLeftColor={borderColor}
      direction="column"
      justifyContent="space-between"
    >
      <Flex direction="column" overflowY="scroll">
        <Flex padding={4} justifyContent="space-between" alignItems="center">
          <Heading noOfLines={1} size="md" textAlign="center">
            Players
          </Heading>

          <CloseButton onClick={onClose} />
        </Flex>

        <PlayerAccordion borderColor={borderColor} />
      </Flex>
    </Flex>
  );
}
