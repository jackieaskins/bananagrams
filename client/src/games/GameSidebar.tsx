import React from 'react';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  Drawer,
  List,
  ListItem,
  ListItemIcon,
} from '@material-ui/core';
import { ExitToApp } from '@material-ui/icons';
import { Link } from 'react-router-dom';

import { useStyles } from '../styles';
import { useGameSidebar } from './GameSidebarState';

const GameSidebar: React.FC = () => {
  const classes = useStyles();
  const {
    leaveGameDialogOpen,
    showLeaveGameDialog,
    handleLeaveGameCancel,
  } = useGameSidebar();

  return (
    <Drawer
      className={classes.drawer}
      classes={{ paper: classes.drawerPaper }}
      variant="permanent"
      anchor="left"
      open
    >
      <List disablePadding>
        <ListItem button onClick={showLeaveGameDialog}>
          <ListItemIcon>
            <ExitToApp />
          </ListItemIcon>
        </ListItem>
      </List>

      <Dialog open={leaveGameDialogOpen} onClose={handleLeaveGameCancel}>
        <DialogContent>
          <DialogContentText>
            Are you sure you want to leave the game?
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleLeaveGameCancel}>No</Button>
          <Button component={Link} to="/" color="primary">
            Yes
          </Button>
        </DialogActions>
      </Dialog>
    </Drawer>
  );
};

export default GameSidebar;
