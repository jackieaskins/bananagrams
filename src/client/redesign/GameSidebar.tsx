import { IconButton } from "@chakra-ui/react";
import { FaBars } from "react-icons/fa6";

export default function GameSidebar(): JSX.Element {
  return (
    <IconButton
      height="100vh"
      borderRadius={0}
      aria-label="Expand game sidebar"
      alignItems="start"
      paddingTop="16px"
      icon={<FaBars />}
    />
  );
}
