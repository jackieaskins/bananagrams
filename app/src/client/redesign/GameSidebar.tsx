import {
  Accordion,
  AccordionButton,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Box,
  CloseButton,
  Flex,
  Heading,
  IconButton,
  Text,
  useColorModeValue,
} from "@chakra-ui/react";
import { FaBars } from "react-icons/fa6";
import PlayerPreview from "./PlayerPreview";
import { useGame } from "@/client/games/GameContext";
import { socket } from "@/client/socket";
import { PlayerStatus } from "@/types/player";

export type GameSidebarProps = {
  expanded: boolean;
  setExpanded: (expanded: boolean) => void;
};

export default function GameSidebar({
  expanded,
  setExpanded,
}: GameSidebarProps): JSX.Element {
  const borderColor = useColorModeValue("gray.400", "gray.700");
  const {
    gameInfo: { players },
  } = useGame();

  if (!expanded) {
    return (
      <IconButton
        height="100vh"
        borderRadius={0}
        aria-label="Expand game sidebar"
        alignItems="start"
        paddingTop="16px"
        icon={<FaBars />}
        onClick={() => setExpanded(true)}
      />
    );
  }

  return (
    <Flex
      height="100vh"
      width="350px"
      borderLeftWidth={1}
      borderLeftColor={borderColor}
      direction="column"
      justifyContent="space-between"
    >
      <Flex direction="column" overflowY="scroll">
        <Flex
          padding={4}
          borderBottomWidth={1}
          borderBottomColor={borderColor}
          justifyContent="space-between"
          alignItems="center"
        >
          <Heading noOfLines={1} size="md" textAlign="center">
            Players
          </Heading>

          <CloseButton onClick={() => setExpanded(false)} />
        </Flex>

        <Accordion allowMultiple overflowY="scroll">
          {players.map(({ userId, username, hand, board, status }, index) => (
            <AccordionItem
              key={userId}
              isDisabled={userId === socket.id}
              borderTop={index === 0 ? 0 : undefined}
              borderBottom={index === players.length - 1 ? 0 : undefined}
              borderColor={borderColor}
            >
              {(isExpanded) => (
                <>
                  <h2>
                    <AccordionButton>
                      <Box as="span" flex="1" textAlign="left">
                        {username}
                      </Box>
                      <AccordionIcon />
                    </AccordionButton>
                  </h2>

                  <AccordionPanel>
                    {isExpanded ? (
                      status === PlayerStatus.SPECTATING ? (
                        <Text textAlign="center">{username} is spectating</Text>
                      ) : (
                        <Flex direction="column">
                          <PlayerPreview
                            board={board}
                            hand={hand}
                            tileSize={16}
                          />
                        </Flex>
                      )
                    ) : null}
                  </AccordionPanel>
                </>
              )}
            </AccordionItem>
          ))}
        </Accordion>
      </Flex>
    </Flex>
  );
}
