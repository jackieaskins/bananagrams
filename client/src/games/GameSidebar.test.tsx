import React from 'react';
import { shallow } from 'enzyme';
import GameSidebar from './GameSidebar';

jest.mock('../styles', () => ({
  useStyles: () => ({
    drawerPaper: 'drawerPaper',
  }),
}));

jest.mock('./GameSidebarState', () => ({
  useGameSidebar: () => ({
    leaveGameDialogOpen: true,
    showLeaveGameDialog: jest.fn().mockName('showLeaveGameDialog'),
    handleLeaveGameCancel: jest.fn().mockName('handleLeaveGameCancel'),
  }),
}));

describe('<GameSidebar />', () => {
  test('renders properly', () => {
    expect(shallow(<GameSidebar />)).toMatchSnapshot();
  });
});
