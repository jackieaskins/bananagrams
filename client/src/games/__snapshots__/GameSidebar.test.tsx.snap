// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameSidebar /> renders properly 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "drawerPaper",
    }
  }
  open={true}
  variant="permanent"
>
  <WithStyles(ForwardRef(List))
    disablePadding={true}
  >
    <WithStyles(ForwardRef(ListItem))
      button={true}
      onClick={[MockFunction showLeaveGameDialog]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <ExitToAppIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Dialog))
    onClose={[MockFunction handleLeaveGameCancel]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))>
        Are you sure you want to leave the game?
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        onClick={[MockFunction handleLeaveGameCancel]}
      >
        No
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Link",
            "propTypes": Object {
              "innerRef": [Function],
              "onClick": [Function],
              "replace": [Function],
              "target": [Function],
              "to": [Function],
            },
            "render": [Function],
          }
        }
        to="/"
      >
        Yes
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</WithStyles(ForwardRef(Drawer))>
`;
